code/game/objects/structures/crates_lockers/closets.dm

```diff
@@ -439,7 +568,135 @@
 
 /obj/structure/closet/proc/tool_interact(obj/item/W, mob/living/user)//returns TRUE if attackBy call shouldn't be continued (because tool was used/closet was of wrong type), FALSE if otherwise
 	. = TRUE
-	if(opened)
+	var/obj/item/card/id/id = null
+	if(!opened && istype(W, /obj/item/airlock_painter))
+		if(!length(paint_jobs))
+			return
+		var/choice = tgui_input_list(user, "Set Closet Paintjob", "Paintjob", paint_jobs)
+		if(isnull(choice))
+			return
+
+		var/obj/item/airlock_painter/painter = W
+		if(!painter.use_paint(user))
+			return
+		var/list/paint_job = paint_jobs[choice]
+		icon_state = paint_job["icon_state"]
+		base_icon_state = icon_state
+		icon_door = paint_job["icon_door"]
+
+		update_appearance()
+
+	else if(can_install_electronics && length(access_choices) && !card_reader_installed && (welded || !can_weld_shut) && istype(W, /obj/item/stock_parts/card_reader))
+		if(broken)
+			balloon_alert(user, "its broken!")
+			return
+
+		if(!secure)
+			balloon_alert(user, "no electronics inside!")
+			return
+
+		if(locked)
+			balloon_alert(user, "unlock first!")
+			return
+
+		user.visible_message(span_notice("[user] is installing a card reader."),
+					span_notice("You begin installing the card reader."))
+
+		if(!do_after(user, 4 SECONDS, target = src))
+			return
+		if(!(secure && !broken && !locked && can_install_electronics && length(access_choices) && !card_reader_installed && (welded || !can_weld_shut)) || !user.transferItemToLoc(W, src))
+			return
```

Comment:
stick these inside a custom proc and then pass it in `do_after` via the `extra_checks` callback.

Reply:
Already did that